video/out/opengl/hwdec_ios.m

```diff
@@ -0,0 +1,295 @@
+}
+
+static bool check_hwdec(struct gl_hwdec *hw)
+{
+    if (hw->gl->es != 300) {
+        MP_ERR(hw, "need OpenGLES 3.0 for CVOpenGLESTextureCacheCreateTextureFromImage()\n");
+        return false;
+    }
+
+    if ([EAGLContext currentContext] == nil) {
+        MP_ERR(hw, "need a current EAGLContext set\n");
+        return false;
+    }
+
+    return true;
+}
+
+static uint32_t get_vt_fmt(struct mp_vt_ctx *vtctx)
+{
+    struct gl_hwdec *hw = vtctx->priv;
+    struct vt_format *f =
+        vt_get_gl_format_from_imgfmt(hw->global->opts->videotoolbox_format);
+    return f ? f->cvpixfmt : (uint32_t)-1;
+}
+
+static int create_hwdec(struct gl_hwdec *hw)
+{
+    if (!check_hwdec(hw))
+        return -1;
+
+    struct priv *p = talloc_zero(hw, struct priv);
+    hw->priv = p;
+
+    CVReturn err = CVOpenGLESTextureCacheCreate(
+        kCFAllocatorDefault,
+        NULL,
+        [EAGLContext currentContext],
+        NULL,
+        &p->gl_texture_cache);
+
+    if (err != noErr) {
+        MP_ERR(hw, "Failure in CVOpenGLESTextureCacheCreate: %d\n", err);
+        return -1;
+    }
+
+    p->vtctx = (struct mp_vt_ctx){
+        .priv = hw,
+        .get_vt_fmt = get_vt_fmt,
+    };
+    p->hwctx = (struct mp_hwdec_ctx){
+        .type = HWDEC_VIDEOTOOLBOX,
+        .download_image = download_image,
+        .ctx = &p->vtctx,
+    };
+    hwdec_devices_add(hw->devs, &p->hwctx);
+
+    return 0;
+}
+
+static int reinit(struct gl_hwdec *hw, struct mp_image_params *params)
+{
+    assert(params->imgfmt == hw->driver->imgfmt);
+
+    struct vt_format *f = vt_get_gl_format_from_imgfmt(params->hw_subfmt);
+    if (!f) {
+        MP_ERR(hw, "Unsupported CVPixelBuffer format.\n");
+        return -1;
+    }
+
+    params->imgfmt = f->imgfmt;
+    params->hw_subfmt = 0;
+    return 0;
+}
+
+static void cleanup_textures(struct gl_hwdec *hw)
+{
+    struct priv *p = hw->priv;
+    int i;
+
+    for (i = 0; i < MP_MAX_PLANES; i++) {
+        if (p->gl_planes[i]) {
+            CFRelease(p->gl_planes[i]);
+            p->gl_planes[i] = NULL;
+        }
+    }
+
+    CVOpenGLESTextureCacheFlush(p->gl_texture_cache, 0);
+}
+
+static int map_frame(struct gl_hwdec *hw, struct mp_image *hw_image,
+                     struct gl_hwdec_frame *out_frame)
+{
+    struct priv *p = hw->priv;
+    GL *gl = hw->gl;
+
+    CVPixelBufferRelease(p->pbuf);
+    p->pbuf = (CVPixelBufferRef)hw_image->planes[3];
+    CVPixelBufferRetain(p->pbuf);
+
+    uint32_t cvpixfmt = CVPixelBufferGetPixelFormatType(p->pbuf);
+    struct vt_format *f = vt_get_gl_format(cvpixfmt);
+    if (!f) {
+        MP_ERR(hw, "CVPixelBuffer has unsupported format type\n");
+        return -1;
+    }
+
+    const bool planar = CVPixelBufferIsPlanar(p->pbuf);
+    const int planes  = CVPixelBufferGetPlaneCount(p->pbuf);
+    assert(planar && planes == f->planes || f->planes == 1);
+
+    cleanup_textures(hw);
+
+    for (int i = 0; i < f->planes; i++) {
+        gl->ActiveTexture(GL_TEXTURE0 + i);
```

Comment:
This line looks pretty wrong and can probably be fully removed. I'm surprised if it doesn't cause other issues. mpv normally leaves GL_TEXTURE0 active, and switches if required (which is basically only when running shaders, because shader textures refer texture units instead of named textures).